### PR TITLE
Visual testing guide

### LINKED ISSUE #1535: Create visual testing guide
based on plugins already listed in http://on.cypress.io/plugins#visual-testing and these tutorials

- Applitools: https://glebbahmutov.com/blog/testing-a-chart/
- Percy: https://glebbahmutov.com/blog/testing-visually/

### PR DESCRIPTION
**Pull request by bahmutov**
- closes #1535 
- [x] it would be nice to include a visual diff image for the ".completed" style change that catches the visual regression

### CODE REVIEW

**Comment by bahmutov** [comment]
@jennifer-shehane I have added Applitools links, so now it would be nice to get this page merged

**Comment by bahmutov** [comment]
@mapsandapps could you give a quick look to this visual testing guide, please? Since you have used it with the services, it would be nice to see if I missed something that developers should know here

File: source/guides/tooling/visual-testing.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+Cypress does NOT check how the page looks though. For example, Cypress does care that the CSS class "completed" grays out the label element and adds a strike-through line.
+
+{% img /img/guides/visual-testing/completed-item.png "Completed item style" %}
+
+Someone could accidentally remove the CSS style ".todo-list li.completed label", changing the application and its usability - and the functional test would never warn you.
+
+Luckily, Cypress gives a stable platform for writing plugins that _can perform visual testing_. Typically such plugins take an image snapshot of the entire application or a specific element, and then compare the image to a previously approved baseline image. If the images are the same (within pixel tolerance), the web application looks the same to the user. If there are differences, then there is some change in the DOM layout, fonts, colors or other visual properties that needs to be investigated.
+
+For example one can use {% url 'cypress-plugin-snapshots' https://github.com/meinaart/cypress-plugin-snapshots %} plugin and catch the following visual regression:
```
[the remark this answers is not in the document]
Comment (bahmutov) [reply source/guides/tooling/visual-testing.md]: absolutely, but think of this page as take-off point for jumping into 3rd party services

File: source/guides/tooling/visual-testing.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+- consider using visual diffing to check individual elements rather than the entire page. This will target the image comparison better to avoid a visual change in the component X from breaking tests for the unrelated components.
+
+- you might want to look at the {% url "Component Testing plugins" plugins %} to load and test the individual components in addition to the regular end-to-end functional and visual tests.
+
+- consider the development costs of implementing the visual testing yourself vs using an external 3rd party provider. Storing, reviewing and analyzing image differences are non-trivial tasks, and they quickly becomes a chore when going with a DIY solution.
+
+## See also
+
+- "Visual Testing" section in the {% url recipes recipes %} page
```
[the remark this answers is not in the document]
Comment (bahmutov) [reply source/guides/tooling/visual-testing.md]: good catch!

**Comment by bahmutov** [comment]
do you think we need more editing here or this is good as v1.0 of the guide?

**Comment by mapsandapps** [comment]
Looks super useful to me. I also found the Applitools and Percy official Cypress integration tutorials to be useful:
* [Percy](https://docs.percy.io/docs/cypress-tutorial)
* [Applitools](https://applitools.com/tutorials/cypress.html#how-it-works)

File: source/guides/tooling/visual-testing.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bahmutov) [reply source/guides/tooling/visual-testing.md]: why is THIS pull request blocked by https://github.com/cypress-io/cypress-monorepo/issues/652 ? isn't the `on` redirect a nice to have, but independent feature?

[the remark this answers is not in the document]
Comment (jennifer-shehane) [reply source/guides/tooling/visual-testing.md]: hmm, I don't know. I just expect the redirects to work before releasing. Having on.cypress.io in the monorepo I think is a bit of a mistake due to it's release cycle though.

**Pull request merged**